### ISSUE TITLE
Windows containers do not use DNS suffix search lists and do not have entries in the hosts file

### ISSUE THREAD

**Issue by bclau** [kubernetes/kubernetes#65016]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

/sig windows

**What happened**:

When deploying Kubernetes pods, the Linux containers will contain the following entries in the ``/etc/resolv.conf`` file:

    nameserver 10.0.0.10
    search e2e-tests-dns-x649p.svc.cluster.local svc.cluster.local cluster.local jqfxe3n5w2rubdex0hknvvd1kf.bx.internal.cloudapp.net
    options ndots:5

and the ``/etc/hosts`` file will contain the following:

    # Kubernetes-managed hosts file.
    127.0.0.1       localhost
    ::1     localhost ip6-localhost ip6-loopback
    fe00::0 ip6-localnet
    fe00::0 ip6-mcastprefix
    fe00::1 ip6-allnodes
    fe00::2 ip6-allrouters
    10.240.0.67     dns-querier-1.dns-test-service.e2e-tests-dns-x649p.svc.cluster.local    dns-querier-1

However, these are not set on Windows containers. First of all, the container's ``C:\Windows\System32\drivers\etc\hosts`` file does not contain any entry. Secondly, the DNS suffix search list is not applied. Only the "Connection-specific DNS suffix" to "svc.cluster.local" (on acs-engine 0.18.0) is set. This means that:

    - kubernetes
    - kubernetes.default
    - kubernetes.default.cluster
    - other pods / services in other namespaces
    
are not all resolvable.

For consistency with the Linux containers, the DNS suffix search list could be applied globally inside the Windows containers:

    $dnsSuffixes = @("namespace.svc.cluster.local", "default.svc.cluster.local", "svc.cluster.local", "cluster.local")
    Set-DnsClientGlobalSetting -SuffixSearchList $dnsSuffixes

**What you expected to happen**:

The following names should be resolvable:

    - kubernetes
    - kubernetes.default
    - kubernetes.default.cluster
    - other pods / services in other namespaces

**How to reproduce it (as minimally and precisely as possible)**:

Create a Windows pod:

        kubectl create namespace e2e-tests-dns-x649p
    kubectl create -f windows_pod.yaml  # https://paste.ubuntu.com/p/ZgfQS9Wgcf/

    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- cmd /S /C type C:\\Windows\\System32\\drivers\\etc\\hosts
    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- cmd /S /C ipconfig /all
    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- cmd /S /C nslookup kubernetes
    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- cmd /S /C nslookup kubernetes.default
    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- cmd /S /C nslookup kubernetes.default.svc
    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- cmd /S /C nslookup dns-test-service
    
    
And a Linux pod for comparison:

    kubectl create namespace e2e-tests-dns-x649p
    kubectl create -f linux_pod.yaml  # https://paste.ubuntu.com/p/kcFcfxYjm3/

    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- cat /etc/hosts
    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- cat /etc/resolv.conf
    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- nslookup kubernetes
    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- nslookup kubernetes.default
    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- nslookup kubernetes.default.svc
    kubectl exec -n e2e-tests-dns-x649p dns-test-2ed58282-6d85-11e8-89cc-000c29738e2c -c querier -- nslookup dns-test-service

**Anything else we need to know?**:

**Environment**: 
- Kubernetes version (use `kubectl version`): 1.10
- Cloud provider or hardware configuration: 
- OS (e.g. from /etc/os-release): Windows Server 1803 (10.0.17134.48)
- Install tools: acs-engine v0.18.0
- Others:

**Comment by PatrickLang** [comment]
This could be a partial fix https://github.com/Azure/azure-container-networking/pull/152

**Comment by PatrickLang** [comment]
@dineshgovindasamy @daschott can you weigh in on how this should be set?

**Comment by saiyan86** [comment]
We should change the default CNI config file to include all these suffixes & change CNI to accommodate these suffixes accordingly. @PatrickLang

**Comment by PatrickLang** [comment]
/remove-lifecycle stale

This is still active and in review #67435 

It's a must fix for DNS conformance tests:

- [sig-network] DNS  should provide DNS for services   [Conformance]
- [sig-network] DNS should provide DNS for the cluster   [Conformance]

**Comment by PatrickLang** [comment]
@bclau - can you update this? We should actually be tracking two separate issues here as the fixes are different:

**Issue 1** - "Windows containers do not use DNS suffix search lists"
That is addressed in #67435, along with related changes to the CNI plugins out of tree

**Issue 2** -  "Windows containers do not have entries in the hosts file"
Can you open a separate issue for that? It can be fixed with a change to the kubelet once Windows Server 2019 has shipped, and once CRI-ContainerD is enabled on Windows. We cannot fix it for Windows Server 1803 or earlier with dockershim.

**Comment by bclau** [comment]
@PatrickLang - indeed, there are basically 2 separate issues here with different solutions. Will do.

Here's the Issue 2's bug report: https://github.com/kubernetes/kubernetes/issues/70189

**Comment by nikopen** [comment]
@bclau let us know about updates on this, code freeze is in about 10 days

**Comment by ylhyh** [comment]
I'm using win-bridge plugin, here is the cni config:
```
{
    "cniVersion": "0.2.0",
    "name": "l2bridge",
    "type": "win-bridge",
    "capabilities": {
        "portMappings": true
    },
    "dns": {
        "Nameservers": [
            "10.176.0.10"
        ],
        "Search": [
            "default.svc.cluster.local",
            "svc.cluster.local",
            "cluster.local"
        ]
    },
    "AdditionalArgs": [
        {
            "Name": "EndpointPolicy",
            "Value": {
                "Type": "OutBoundNAT",
                "ExceptionList": [
                    "10.160.0.0/12",
                    "10.176.0.0/12"
                ]
            }
        },
        {
            "Name": "EndpointPolicy",
            "Value": {
                "Type": "ROUTE",
                "DestinationPrefix": "10.176.0.0/12",
                "NeedEncap": true
            }
        }
    ]
}
```

By executing ipconfig /all in the Windows Pod, I can see the pod retrieved correct DNS settings (server and suffix search list):
```
C:\>ipconfig /all
                                                                                                                                                                                      
Windows IP Configuration

   Host Name . . . . . . . . . . . . : netfx-test-98f577b9f-tm66z
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : default.svc.cluster.local
                                       svc.cluster.local
                                       cluster.local

Ethernet adapter vEthernet (8868b953a9db6d04d78548c27373e373e0ba9f6f7e8fce6778f94d5cde487a29_l2bridge): 

   Connection-specific DNS Suffix  . : default.svc.cluster.local
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #18
   Physical Address. . . . . . . . . : 00-15-5D-64-D8-69
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::2480:8337:bb94:61a4%29(Preferred)
   IPv4 Address. . . . . . . . . . . : 10.160.12.84(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 10.160.12.1
   DNS Servers . . . . . . . . . . . : 10.176.0.10
   NetBIOS over Tcpip. . . . . . . . : Disabled
   Connection-specific DNS Suffix Search List :
                                       default.svc.cluster.local
                                       svc.cluster.local
                                       cluster.local
```
But the DNS suffix Search list is not work:
```
C:\>nslookup kubernetes.default
Server:  kube-dns.kube-system.svc.cluster.local
Address:  10.176.0.10

*** kube-dns.kube-system.svc.cluster.local can't find kubernetes.default: Non-existent domain
```

Execute "Get-DnsClientGlobalSetting" in powershell, UseSuffixSearchList is False:
```powershell
PS C:\> Get-DnsClientGlobalSetting
                                                                                                                                                                       
UseSuffixSearchList : False
SuffixSearchList    : {}
UseDevolution       : True
DevolutionLevel     : 0
```

If I manually set Suffix Search List in the Pod by execute "Set-DnsClientGlobalSetting  -SuffixSearchList ...", DNS suffix search list will work:
```powershell
PS C:\> Set-DnsClientGlobalSetting -SuffixSearchList @("default.svc.cluster.local", "svc.cluster.local", "cluster.local")
PS C:\> Get-DnsClientGlobalSetting

UseSuffixSearchList : True
SuffixSearchList    : {default.svc.cluster.local, svc.cluster.local, cluster.local}
UseDevolution       : True
DevolutionLevel     : 0

PS C:\> nslookup kubernetes                                                                                                                                                                                     
Server:  kube-dns.kube-system.svc.cluster.local                                                                                                                                                                 
Address:  10.176.0.10                                                                                                                                                                                           
                                                                                                                                                                                                                
Name:    kubernetes.default.svc.cluster.local                                                                                                                                                                   
Address:  10.176.0.1                                                                                                                                                                                            
                                                                                                                                                                                                                
PS C:\> nslookup kubernetes.default
Server:  kube-dns.kube-system.svc.cluster.local
Address:  10.176.0.10

Name:    kubernetes.default.svc.cluster.local
Address:  10.176.0.1
```
My environment:
- kubelet 1.17.0
- Windows Node OS: 10.0.17763.973
- Windows Container OS: 10.0.17763.973
- CNI plugin 0.8.5

@feiskyer can you give some input? how to make DNS suffix search list work automatically?

**Comment by feiskyer** [comment]
@marosset please have a look at the above issues

**Comment by ksubrmnn** [comment]
@dineshgovindasamy @JocelynBerrendonner PTAL

**Comment by claudiubelu** [comment]
@ylhyh Interesting find. There are a few things that I would like to correct to what you've written there, but for the most part, it is worth implementing.

Let's see. The thing that is not quite correct is that after setting the DNS Suffix Search List with ``Set-DnsClientGlobalSetting`` you'll be able to completely resolve Partially Qualified Domain Names. Yes, I've checked, ``nslookup kubernetes.default`` works afterwards, but that's pretty much it. It was not working for anything else I've tried: ping, curl, wget, not even Resolve-DnsName.

Now, why is this still helpfull: it will allow us to correctly resolve a subset of PQDNs, in particular, hostname PQDNs (don't know how to be more specific than this). Basically, it will allow us to have the following functionality:

    PS C:\> Get-DnsClientGlobalSetting

    UseSuffixSearchList : False
    SuffixSearchList    : {}
    UseDevolution       : True
    DevolutionLevel     : 0

    PS C:\> ping kubernetes
    Ping request could not find host kubernetes. Please check the name and try again.
    PS C:\> nslookup kubernetes
    Server:  kube-dns.kube-system.svc.cluster.local
    Address:  10.0.0.10

    *** kube-dns.kube-system.svc.cluster.local can't find kubernetes: Non-existent domain

    PS C:\> Resolve-DnsName kubernetes
    Resolve-DnsName : kubernetes : DNS name does not exist
    At line:1 char:1
    + Resolve-DnsName kubernetes
    + ~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : ResourceUnavailable: (kubernetes:String) [Resolve-DnsName], Win32Exception
        + FullyQualifiedErrorId : DNS_ERROR_RCODE_NAME_ERROR,Microsoft.DnsClient.Commands.ResolveDnsName

    PS C:\> curl "https://kubernetes:443"
    curl : The remote name could not be resolved: 'kubernetes'
    At line:1 char:1
    + curl "https://kubernetes:443"
    + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], W
       ebException
        + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand

    PS C:\> curl.exe "https://kubernetes:443"
    curl: (6) Could not resolve host: kubernetes

    # After setting the DNS suffix list.
    PS C:\> Set-DnsClientGlobalSetting -SuffixSearchList @("default.svc.cluster.local", "svc.cluster.local", "cluster.l
    ocal")
    PS C:\> ping kubernetes

    Pinging kubernetes.default.svc.cluster.local [10.0.0.1] with 32 bytes of data:
    Control-C
    PS C:\> nslookup kubernetes
    Server:  kube-dns.kube-system.svc.cluster.local
    Address:  10.0.0.10

    Name:    kubernetes.default.svc.cluster.local
    Address:  10.0.0.1

    PS C:\> Resolve-DnsName kubernetes

    Name                                           Type   TTL   Section    IPAddress
    ----                                           ----   ---   -------    ---------
    kubernetes.default.svc.cluster.local           A      5     Answer     10.0.0.1

    # this still failed for me, but it resolved the name:
    PS C:\> $KUBE_TOKEN=cat C:\var\run\secrets\kubernetes.io\serviceaccount\token
    PS C:\> $hostname = hostname
    PS C:\> curl.exe -sSk -H "Authorization: Bearer $KUBE_TOKEN" "https://kubernetes:443"
    {
      "kind": "Status",
      "apiVersion": "v1",
      "metadata": {

      },
      "status": "Failure",
      "message": "forbidden: User \"system:serviceaccount:test-windows-containers2:default\" cannot get path \"/\"",
      "reason": "Forbidden",
      "details": {

      },
      "code": 403
}

So, as you can see, it's still useful for the above cases.

**Comment by ylhyh** [comment]
@claudiubelu, I think the "403 Forbidden" issue is caused by default RBAC setting.

As your reminder, I found another thing interesting:

1. Set Suffix Search List to **["default.svc.cluster.local", "svc.cluster.local", "cluster.local"]** by executing by Set-DnsClientGlobalSetting:
  - for nslookup:
    - kubernetes - WORK
    - kubernetes.default - WORK
    - kubernetes.default.svc - WORK
    - kubernetes.default.svc.cluster.local - WORK

  - for Powershell's curl:
    - kubernetes - WORK
    - kubernetes.default - NOT WORK
    - kubernetes.default.svc - NOT WORK
    - kubernetes.default.svc.cluster.local - WORK

  - for curl.exe:
    - kubernetes - WORK
    - kubernetes.default - NOT WORK
    - kubernetes.default.svc - NOT WORK
    - kubernetes.default.svc.cluster.local - WORK

2. Set Suffix Search List to **["svc.cluster.local", "cluster.local"]** by executing by Set-DnsClientGlobalSetting:
  - for nslookup:
    - kubernetes - NOT WORK
    - kubernetes.default - WORK
    - kubernetes.default.svc - WORK
    - kubernetes.default.svc.cluster.local - WORK

  - for Powershell's curl:
    - kubernetes - NOT WORK
    - kubernetes.default - NOT WORK
    - kubernetes.default.svc - NOT WORK
    - kubernetes.default.svc.cluster.local - WORK

  - for curl.exe:
    - kubernetes - NOT WORK
    - kubernetes.default - NOT WORK
    - kubernetes.default.svc - NOT WORK
    - kubernetes.default.svc.cluster.local - WORK

3. Set Suffix Search List to **["cluster.local"]** by executing by Set-DnsClientGlobalSetting:
  - for nslookup:
    - kubernetes - NOT WORK
    - kubernetes.default - NOT WORK
    - kubernetes.default.svc - WORK
    - kubernetes.default.svc.cluster.local - WORK

  - for Powershell's curl:
    - kubernetes - NOT WORK
    - kubernetes.default - NOT WORK
    - kubernetes.default.svc - NOT WORK
    - kubernetes.default.svc.cluster.local - WORK

  - for curl.exe:
    - kubernetes - NOT WORK
    - kubernetes.default - NOT WORK
    - kubernetes.default.svc - NOT WORK
    - kubernetes.default.svc.cluster.local - WORK

It seems "DNS Suffix Search List" only works for nslookup? I'm get confused

**Comment by claudiubelu** [comment]
> 
> @claudiubelu, I think the "403 Forbidden" issue is caused by default RBAC setting.
> 

Indeed, nonetheless, the 403 is given by the Kubernetes API itself, which means that the name was resolved and connected to, which is why we got the server response.

> 
> It seems "DNS Suffix Search List" only works for nslookup? I'm get confused

So, one small note on the experiments you've done there: the DNS Suffix Search List is not useful for Fully Qualified Domain Names like ``kubernetes.default.svc.cluster.local``. The purpose of those DNS Suffix Search List is to be able to resolve a Partially Qualified Domain Names by appending one of the suffixes, which will result in a FQDN (for example: ``kubernetes.default`` + ``svc.cluster.local`` = ``kubernetes.default.svc.cluster.local``, or ``kubernetes`` + ``default.svc.cluster.local`` = ``kubernetes.default.svc.cluster.local``).

So, basically, removing ``default.svc.cluster.local`` from the DNS suffix list will result in ``kubernetes`` no longer being resolvable: there is no suffix that will get it to the proper FQDN.

> 
> It seems "DNS Suffix Search List" only works for nslookup? I'm get confused

And to answer your question, yes, it seems that only ``nslookup`` can resolve **all** PQDNs.  In all other cases, it seems that ``DNS Suffix Search List`` is still useful for Partially Qualified Domain Names that do not have any dots in them (like ``kubernetes``, or any ``some-service-name`` that you've added to your Kubernetes).

**Comment by ylhyh** [comment]
@claudiubelu

`ClusterFirstWithHostNet is not supported for DNS. Windows treats all names with a ‘.’ as a FQDN and skips PQDN resolution`

Described here: https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/
